packages/mongodb-mql-engines/src/main/kotlin/com/mongodb/jbplugin/indexing/IndexQueryFieldUsage.kt

```diff
@@ -0,0 +1,166 @@
+package com.mongodb.jbplugin.indexing
+
+import com.mongodb.jbplugin.indexing.IndexAnalyzer.SortDirection
+import com.mongodb.jbplugin.mql.BsonAny
+import com.mongodb.jbplugin.mql.BsonType
+import com.mongodb.jbplugin.mql.CollectionSchema
+import com.mongodb.jbplugin.mql.Node
+import com.mongodb.jbplugin.mql.components.HasFieldReference
+import com.mongodb.jbplugin.mql.components.Name
+import com.mongodb.jbplugin.mql.components.QueryRole
+import com.mongodb.jbplugin.mql.parser.NoConditionFulfilled
+import com.mongodb.jbplugin.mql.parser.anyError
+import com.mongodb.jbplugin.mql.parser.components.aggregationStages
+import com.mongodb.jbplugin.mql.parser.components.allNodesWithSchemaFieldReferences
+import com.mongodb.jbplugin.mql.parser.components.extractOperation
+import com.mongodb.jbplugin.mql.parser.components.extractValueReferencesRelevantForIndexing
+import com.mongodb.jbplugin.mql.parser.components.fieldReference
+import com.mongodb.jbplugin.mql.parser.components.hasName
+import com.mongodb.jbplugin.mql.parser.filter
+import com.mongodb.jbplugin.mql.parser.first
+import com.mongodb.jbplugin.mql.parser.flatMap
+import com.mongodb.jbplugin.mql.parser.map
+import com.mongodb.jbplugin.mql.parser.mapMany
+import com.mongodb.jbplugin.mql.parser.matches
+import com.mongodb.jbplugin.mql.parser.nth
+import com.mongodb.jbplugin.mql.parser.parse
+import com.mongodb.jbplugin.mql.parser.requireNonNull
+import com.mongodb.jbplugin.mql.parser.zip
+
+internal data class IndexQueryFieldUsage<S>(
+    val source: S,
+    val fieldName: String,
+    val fieldType: BsonType,
+    val value: Any?,
+    val valueType: BsonType,
+    val role: QueryRole,
+    val selectivity: Double?,
+    val sortDirection: SortDirection,
+    val reference: Node<S>?,
+    val operation: Name,
+) {
+    companion object {
+        fun <S> byRoleSelectivityAndCardinality() = Comparator<IndexQueryFieldUsage<S>> { a, b ->
+            val roleComparison = a.role.ordinal.compareTo(b.role.ordinal)
+            if (roleComparison != 0) return@Comparator roleComparison
+
+            if (a.selectivity != null && b.selectivity != null) {
+                val selectivityComparison = b.selectivity.compareTo(a.selectivity)
+                if (selectivityComparison != 0) return@Comparator selectivityComparison
+            }
+
+            val cardinalityComparison = if (a.role == QueryRole.SORT && b.role == QueryRole.SORT) {
+                a.fieldType.cardinality.compareTo(b.fieldType.cardinality)
+            } else {
+                a.valueType.cardinality.compareTo(b.valueType.cardinality)
+            }
+
+            if (cardinalityComparison != 0) {
+                return@Comparator cardinalityComparison
+            }
+
+            return@Comparator a.fieldName.compareTo(b.fieldName)
```

Comment:
This ensures that for situations where 2 fields are identical in role, selectivity and cardinality, we still have a consistent index, by sorting all the fields that are equally by name.